File: src/quest_editor/scripting/assembly_worker.ts
import {
    AssemblyChangeInput,
    AssemblySettingsChangeInput,
    AssemblyWorkerInput,
    DefinitionInput,
    DefinitionOutput,
    InputMessageType,
    NewObjectCodeOutput,
    OutputMessageType,
    SignatureHelpInput,
    SignatureHelpOutput,
} from "./assembly_worker_messages";
import { assemble, AssemblySettings } from "./assembly";
import { AsmToken, Segment, SegmentType } from "./instructions";
import { Kind, OP_BB_MAP_DESIGNATE, Opcode, OPCODES_BY_MNEMONIC } from "./opcodes";
import { AssemblyLexer, IdentToken, TokenType } from "./AssemblyLexer";

const ctx: Worker = self as any;

let lines: string[] = [];
let object_code: Segment[] = [];
const line_no_to_instructions: {
    segment_index: number;
    instruction_indices: number[];
}[] = [];
const label_to_segment_cache: Map<number, Segment | null> = new Map();

const messages: AssemblyWorkerInput[] = [];
let timeout: any;

const assembly_settings: AssemblySettings = {
    manual_stack: false,
};

ctx.onmessage = (e: MessageEvent) => {
    messages.push(e.data);

    if (!timeout) {
        process_messages();

        timeout = setTimeout(() => {
            timeout = undefined;
            process_messages();
        }, 100);
    }
};

function process_messages(): void {
    if (messages.length === 0) return;

    for (const message of messages.splice(0, messages.length)) {
        switch (message.type) {
            case InputMessageType.NewAssembly:
                lines = message.assembly;
                assemble_and_send();
                break;
            case InputMessageType.AssemblyChange:
                assembly_change(message);
                break;
            case InputMessageType.SignatureHelp:
                signature_help(message);
                break;
            case InputMessageType.Definition:
                definition(message);
                break;
            case InputMessageType.SettingsChange:
                settings_change(message);
                break;
        }
    }
}

function assembly_change(message: AssemblyChangeInput): void {
    for (const change of message.changes) {
        const { start_line_no, end_line_no, start_col, end_col, new_text } = change;
        const lines_changed = end_line_no - start_line_no + 1;
        const new_lines = new_text.split("\n");

        if (lines_changed === 1) {
            replace_line_part(start_line_no, start_col, end_col, new_lines);
        } else if (new_lines.length === 1) {
            replace_lines_and_merge_line_parts(
                start_line_no,
                end_line_no,
                start_col,
                end_col,
                new_lines[0],
            );
        } else {
            // Keep the left part of the first changed line.
            replace_line_part_right(start_line_no, start_col, new_lines[0]);

            // Keep the right part of the last changed line.
            replace_line_part_left(end_line_no, end_col, new_lines[new_lines.length - 1]);

            // Replace all the lines in between.
            // It's important that we do this last.
            replace_lines(
                start_line_no + 1,
                end_line_no - 1,
                new_lines.slice(1, new_lines.length - 1),
            );
        }
    }

    assemble_and_send();
}

// Hacky way of providing parameter hints.
// We just tokenize the current line and look for the first identifier and check whether it's a valid opcode.
// TODO: make use of new meta information in IR.
function signature_help(message: SignatureHelpInput): void {
    let opcode: Opcode | undefined;
    let active_param = -1;

    if (message.line_no <= lines.length) {
        const line = lines[message.line_no - 1];
        const lexer = new AssemblyLexer();
        const tokens = lexer.tokenize_line(line);
        const ident = tokens.find(t => t.type === TokenType.Ident) as IdentToken | undefined;

        if (ident) {
            opcode = OPCODES_BY_MNEMONIC.get(ident.value);

            if (opcode) {
                for (const token of tokens) {
                    if (token.col + token.len > message.col) {
                        break;
                    } else if (token.type === TokenType.Ident && active_param === -1) {
                        active_param = 0;
                    } else if (token.type === TokenType.ArgSeparator) {
                        active_param++;
                    }
                }
            }
        }
    }

    const response: SignatureHelpOutput = {
        type: OutputMessageType.SignatureHelp,
        id: message.id,
        opcode,
        active_param,
    };
    ctx.postMessage(response);
}

function definition(message: DefinitionInput): void {
    const label = get_label_reference_at(message.line_no, message.col);
    let asm: AsmToken | undefined;

    if (label != undefined) {
        const segment = get_segment_by_label(label);

        if (segment) {
            const index = segment.labels.indexOf(label);

            if (index !== -1) {
                asm = segment.asm.labels[index];
            }
        }
    }

    const response: DefinitionOutput = {
        type: OutputMessageType.Definition,
        id: message.id,
        ...asm,
    };
    ctx.postMessage(response);
}

/**
 * Apply changes to settings.
 */
function settings_change(message: AssemblySettingsChangeInput): void {
    if (message.settings.manual_stack != undefined) {
        assembly_settings.manual_stack = message.settings.manual_stack;
    }
}

function assemble_and_send(): void {
    const assembler_result = assemble(lines, assembly_settings.manual_stack);

    object_code = assembler_result.object_code;
    label_to_segment_cache.clear();
    line_no_to_instructions.splice(0, Infinity);

    const map_designations = new Map<number, number>();

    for (let i = 0; i < object_code.length; i++) {
        const segment = object_code[i];

        if (segment.type === SegmentType.Instructions) {
            // Set map designations.
            if (segment.labels.includes(0)) {
                for (const inst of segment.instructions) {
                    if (inst.opcode.code === OP_BB_MAP_DESIGNATE.code) {
                        map_designations.set(inst.args[0].value, inst.args[2].value);
                    }
                }
            }

            // Index instructions by text position.
            for (let j = 0; j < segment.instructions.length; j++) {
                const ins = segment.instructions[j];

                if (ins.asm) {
                    if (ins.asm.mnemonic) {
                        add_index(ins.asm.mnemonic.line_no, i, j);
                    }

                    for (const arg_asm of ins.asm.args) {
                        add_index(arg_asm.line_no, i, j);
                    }
                }
            }
        }
    }

    const response: NewObjectCodeOutput = {
        type: OutputMessageType.NewObjectCode,
        map_designations,
        ...assembler_result,
    };
    ctx.postMessage(response);
}

function replace_line_part(
    line_no: number,
    start_col: number,
    end_col: number,
    new_line_parts: string[],
): void {
    const line = lines[line_no - 1];
    // We keep the parts of the line that weren't affected by the edit.
    const line_start = line.slice(0, start_col - 1);
    const line_end = line.slice(end_col - 1);

    if (new_line_parts.length === 1) {
        lines.splice(line_no - 1, 1, line_start + new_line_parts[0] + line_end);
    } else {
        lines.splice(
            line_no - 1,
            1,
            line_start + new_line_parts[0],
            ...new_line_parts.slice(1, new_line_parts.length - 1),
            new_line_parts[new_line_parts.length - 1] + line_end,
        );
    }
}

function replace_line_part_left(line_no: number, end_col: number, new_line_part: string): void {
    lines.splice(line_no - 1, 1, new_line_part + lines[line_no - 1].slice(end_col - 1));
}

function replace_line_part_right(line_no: number, start_col: number, new_line_part: string): void {
    lines.splice(line_no - 1, 1, lines[line_no - 1].slice(0, start_col - 1) + new_line_part);
}

function replace_lines(start_line_no: number, end_line_no: number, new_lines: string[]): void {
    lines.splice(start_line_no - 1, end_line_no - start_line_no + 1, ...new_lines);
}

function replace_lines_and_merge_line_parts(
    start_line_no: number,
    end_line_no: number,
    start_col: number,
    end_col: number,
    new_line_part: string,
): void {
    const start_line = lines[start_line_no - 1];
    const end_line = lines[end_line_no - 1];
    // We keep the parts of the lines that weren't affected by the edit.
    const start_line_start = start_line.slice(0, start_col - 1);
    const end_line_end = end_line.slice(end_col - 1);

    lines.splice(
        start_line_no - 1,
        end_line_no - start_line_no + 1,
        start_line_start + new_line_part + end_line_end,
    );
}

// TODO: make the code work with stack-based instructions
function get_label_reference_at(line_no: number, col: number): number | undefined {
    const handle = line_no_to_instructions[line_no];
    if (!handle) return undefined;

    const segment = object_code[handle.segment_index];
    if (!segment || segment.type !== SegmentType.Instructions) return undefined;

    for (const index of handle.instruction_indices) {
        const ins = segment.instructions[index];

        if (ins && ins.asm) {
            const params = ins.opcode.params;

            for (let i = 0; i < ins.asm.args.length; i++) {
                const param = i < params.length ? params[i] : params[params.length - 1];
                const arg_asm = ins.asm.args[i];

                if (
                    (param.type.kind === Kind.ILabel ||
                        param.type.kind === Kind.DLabel ||
                        param.type.kind === Kind.SLabel ||
                        param.type.kind === Kind.ILabelVar) &&
                    position_inside(line_no, col, arg_asm)
                ) {
                    return ins.args[i].value;
                }
            }

            for (let i = 0; i < ins.asm.stack_args.length; i++) {
                const param = i < params.length ? params[i] : params[params.length - 1];
                const arg_asm = ins.asm.stack_args[i];

                if (
                    (param.type.kind === Kind.ILabel ||
                        param.type.kind === Kind.DLabel ||
                        param.type.kind === Kind.SLabel ||
                        param.type.kind === Kind.ILabelVar) &&
                    position_inside(line_no, col, arg_asm)
                ) {
                    return arg_asm.value;
                }
            }
        }
    }

    return undefined;
}

function get_segment_by_label(label: number): Segment | undefined {
    let segment = label_to_segment_cache.get(label);

    // Strict comparison because null has special meaning.
    if (segment === undefined) {
        segment = null;

        for (const seg of object_code) {
            if (seg.labels.includes(label)) {
                segment = seg;
                break;
            }
        }

        label_to_segment_cache.set(label, segment);
    }

    return segment || undefined;
}

function add_index(line_no: number, segment_index: number, instruction_index: number): void {
    let handle = line_no_to_instructions[line_no];

    if (!handle) {
        handle = { segment_index, instruction_indices: [] };
        line_no_to_instructions[line_no] = handle;
    }

    handle.instruction_indices.push(instruction_index);
}

function position_inside(line_no: number, col: number, asm?: AsmToken): boolean {
    if (asm) {
        return line_no === asm.line_no && col >= asm.col && col <= asm.col + asm.len;
    } else {
        return false;
    }
}

File: web/assembly-worker/src/main/kotlin/world/phantasmal/web/assemblyWorker/AssemblyWorker.kt
package world.phantasmal.web.assemblyWorker

import mu.KotlinLogging
import world.phantasmal.core.*
import world.phantasmal.lib.asm.*
import world.phantasmal.lib.asm.dataFlowAnalysis.ControlFlowGraph
import world.phantasmal.lib.asm.dataFlowAnalysis.getMapDesignations
import world.phantasmal.lib.asm.dataFlowAnalysis.getStackValue
import world.phantasmal.web.shared.Throttle
import world.phantasmal.web.shared.messages.*
import kotlin.math.min
import kotlin.time.measureTime
import world.phantasmal.lib.asm.AssemblyProblem as AssemblerAssemblyProblem

private val logger = KotlinLogging.logger {}

class AssemblyWorker(private val sendMessage: (ServerMessage) -> Unit) {
    private val messageQueue: MutableList<ClientMessage> = mutableListOf()
    private val messageProcessingThrottle = Throttle(wait = 100)
    private val tokenizer = LineTokenizer()

    // User input.
    private var inlineStackArgs: Boolean = true
    private val asm: JsArray<String> = jsArrayOf()

    // Output.
    private var bytecodeIr = BytecodeIr(emptyList())
    private var problems: List<AssemblyProblem>? = null

    // Derived data.
    private var _cfg: ControlFlowGraph? = null
    private val cfg: ControlFlowGraph
        get() {
            if (_cfg == null) _cfg = ControlFlowGraph.create(bytecodeIr)
            return _cfg!!
        }

    private var mapDesignations: Map<Int, Int>? = null

    fun receiveMessage(message: ClientMessage) {
        messageQueue.add(message)
        messageProcessingThrottle(::processMessages)
    }

    private fun processMessages() {
        // Split messages into ASM changes and other messages. Remove useless/duplicate
        // notifications.
        val asmChanges = mutableListOf<ClientNotification>()
        val otherMessages = mutableListOf<ClientMessage>()

        for (message in messageQueue) {
            when (message) {
                is ClientNotification.SetAsm -> {
                    // All previous ASM change messages can be discarded when the entire ASM has
                    // changed.
                    asmChanges.clear()
                    asmChanges.add(message)
                }

                is ClientNotification.UpdateAsm ->
                    asmChanges.add(message)

                else ->
                    otherMessages.add(message)
            }
        }

        messageQueue.clear()

        // Process ASM changes first.
        processAsmChanges(asmChanges)
        otherMessages.forEach(::processMessage)
    }

    private fun processAsmChanges(messages: List<ClientNotification>) {
        if (messages.isNotEmpty()) {
            val time = measureTime {
                for (message in messages) {
                    when (message) {
                        is ClientNotification.SetAsm ->
                            setAsm(message.asm, message.inlineStackArgs)

                        is ClientNotification.UpdateAsm ->
                            updateAsm(message.changes)
                    }
                }

                processAsm()
            }

            logger.trace {
                "Processed ${messages.size} assembly changes in ${time.inMilliseconds}ms."
            }
        }
    }

    private fun processMessage(message: ClientMessage) {
        val time = measureTime {
            when (message) {
                is ClientNotification.SetAsm,
                is ClientNotification.UpdateAsm ->
                    logger.error { "Unexpected ${message::class.simpleName}." }

                is Request.GetCompletions ->
                    getCompletions(message.id, message.lineNo, message.col)

                is Request.GetSignatureHelp ->
                    getSignatureHelp(message.id, message.lineNo, message.col)

                is Request.GetHover ->
                    getHover(message.id, message.lineNo, message.col)

                is Request.GetDefinition ->
                    getDefinition(message.id, message.lineNo, message.col)
            }
        }

        logger.trace { "Processed ${message::class.simpleName} in ${time.inMilliseconds}ms." }
    }

    private fun setAsm(asm: List<String>, inlineStackArgs: Boolean) {
        this.inlineStackArgs = inlineStackArgs
        this.asm.splice(0, this.asm.length, *asm.toTypedArray())
        mapDesignations = null
    }

    private fun updateAsm(changes: List<AsmChange>) {
        for (change in changes) {
            val (startLineNo, startCol, endLineNo, endCol) = change.range
            val linesChanged = endLineNo - startLineNo + 1
            val newLines = change.newAsm.split("\n").toJsArray()

            when {
                linesChanged == 1 -> {
                    replaceLinePart(startLineNo, startCol, endCol, newLines)
                }

                newLines.length == 1 -> {
                    replaceLinesAndMergeLineParts(
                        startLineNo,
                        endLineNo,
                        startCol,
                        endCol,
                        newLines[0],
                    )
                }

                else -> {
                    // Keep the left part of the first changed line.
                    replaceLinePartRight(startLineNo, startCol, newLines[0])

                    // Keep the right part of the last changed line.
                    replaceLinePartLeft(endLineNo, endCol, newLines[newLines.length - 1])

                    // Replace all the lines in between.
                    // It's important that we do this last.
                    replaceLines(
                        startLineNo + 1,
                        endLineNo - 1,
                        newLines.slice(1, newLines.length - 1),
                    )
                }
            }
        }
    }

    private fun replaceLinePart(
        lineNo: Int,
        startCol: Int,
        endCol: Int,
        newLineParts: JsArray<String>,
    ) {
        val line = asm[lineNo - 1]
        // We keep the parts of the line that weren't affected by the edit.
        val lineStart = line.substring(0, startCol - 1)
        val lineEnd = line.substring(endCol - 1)

        if (newLineParts.length == 1) {
            asm[lineNo - 1] = lineStart + newLineParts[0] + lineEnd
        } else {
            asm.splice(
                lineNo - 1,
                1,
                lineStart + newLineParts[0],
                *newLineParts.slice(1, newLineParts.length - 1).asArray(),
                newLineParts[newLineParts.length - 1] + lineEnd,
            )
        }
    }

    private fun replaceLinePartLeft(lineNo: Int, endCol: Int, newLinePart: String) {
        asm[lineNo - 1] = newLinePart + asm[lineNo - 1].substring(endCol - 1)
    }

    private fun replaceLinePartRight(lineNo: Int, startCol: Int, newLinePart: String) {
        asm[lineNo - 1] = asm[lineNo - 1].substring(0, startCol - 1) + newLinePart
    }

    private fun replaceLines(startLineNo: Int, endLineNo: Int, newLines: JsArray<String>) {
        asm.splice(startLineNo - 1, endLineNo - startLineNo + 1, *newLines.asArray())
    }

    private fun replaceLinesAndMergeLineParts(
        startLineNo: Int,
        endLineNo: Int,
        startCol: Int,
        endCol: Int,
        newLinePart: String,
    ) {
        val startLine = asm[startLineNo - 1]
        val endLine = asm[endLineNo - 1]
        // We keep the parts of the lines that weren't affected by the edit.
        val startLineStart = startLine.substring(0, startCol - 1)
        val endLineEnd = endLine.substring(endCol - 1)

        asm.splice(
            startLineNo - 1,
            endLineNo - startLineNo + 1,
            startLineStart + newLinePart + endLineEnd,
        )
    }

    private fun processAsm() {
        _cfg = null

        val assemblyResult = assemble(asm.asArray().toList(), inlineStackArgs)

        @Suppress("UNCHECKED_CAST")
        val problems = (assemblyResult.problems as List<AssemblerAssemblyProblem>).map {
            AssemblyProblem(it.severity, it.uiMessage, it.lineNo, it.col, it.len)
        }

        if (problems != this.problems) {
            this.problems = problems
            sendMessage(ServerNotification.Problems(problems))
        }

        if (assemblyResult is Success) {
            bytecodeIr = assemblyResult.value

            val instructionSegments = bytecodeIr.instructionSegments()

            instructionSegments.find { 0 in it.labels }?.let { label0Segment ->
                val designations = getMapDesignations(label0Segment) { cfg }

                if (designations != mapDesignations) {
                    mapDesignations = designations
                    sendMessage(
                        ServerNotification.MapDesignations(
                            designations
                        )
                    )
                }
            }
        }
    }

    private fun getCompletions(requestId: Int, lineNo: Int, col: Int) {
        val text = getLine(lineNo)?.take(col)?.trim()?.toLowerCase() ?: ""

        val completions: List<CompletionItem> = when {
            KEYWORD_REGEX.matches(text) -> KEYWORD_SUGGESTIONS

            INSTRUCTION_REGEX.matches(text) -> {
                val suggestions = INSTRUCTION_SUGGESTIONS.asSequence()
                val startingWith = suggestions.filter { it.label.startsWith(text) }
                val containing = suggestions.filter { it.label.contains(text) }

                (startingWith + containing)
                    .take(20)
                    .toList()
            }

            else -> emptyList()
        }

        sendMessage(Response.GetCompletions(requestId, completions))
    }

    private fun getSignatureHelp(requestId: Int, lineNo: Int, col: Int) {
        sendMessage(Response.GetSignatureHelp(requestId, signatureHelp(lineNo, col)))
    }

    private fun signatureHelp(lineNo: Int, col: Int): SignatureHelp? {
        // Hacky way of providing parameter hints.
        // We just tokenize the current line and look for the first identifier and check whether
        // it's a valid opcode.
        var signature: Signature? = null
        var activeParam = -1

        getLine(lineNo)?.let { text ->
            tokenizer.tokenize(text)

            while (tokenizer.nextToken()) {
                if (tokenizer.type === Token.Ident) {
                    mnemonicToOpcode(tokenizer.strValue)?.let { opcode ->
                        signature = getSignature(opcode)
                    }
                }

                if (tokenizer.col + tokenizer.len > col) {
                    break
                } else if (tokenizer.type === Token.Ident && activeParam == -1) {
                    activeParam = 0
                } else if (tokenizer.type === Token.ArgSeparator) {
                    activeParam++
                }
            }
        }

        return signature?.let { sig ->
            SignatureHelp(
                signature = sig,
                activeParameter = activeParam,
            )
        }
    }

    private fun getSignature(opcode: Opcode): Signature {
        var signature = opcode.mnemonic + " "
        val params = mutableListOf<Parameter>()
        var first = true

        for (param in opcode.params) {
            if (first) {
                first = false
            } else {
                signature += ", "
            }

            val paramTypeStr = when (param.type) {
                ByteType -> "Byte"
                ShortType -> "Short"
                IntType -> "Int"
                FloatType -> "Float"
                ILabelType -> "&Function"
                DLabelType -> "&Data"
                SLabelType -> "&String"
                ILabelVarType -> "...&Function"
                StringType -> "String"
                RegRefType, is RegTupRefType -> "Register"
                RegRefVarType -> "...Register"
                PointerType -> "Pointer"
                else -> "Any"
            }

            params.add(
                Parameter(
                    labelStart = signature.length,
                    labelEnd = signature.length + paramTypeStr.length,
                    documentation = param.doc,
                )
            )

            signature += paramTypeStr
        }

        return Signature(
            label = signature,
            documentation = opcode.doc,
            parameters = params,
        )
    }

    private fun getHover(requestId: Int, lineNo: Int, col: Int) {
        val hover = signatureHelp(lineNo, col)?.let { help ->
            val sig = help.signature
            val param = sig.parameters.getOrNull(help.activeParameter)

            val contents = mutableListOf<String>()

            // Instruction signature. Parameter highlighted if possible.
            contents.add(
                if (param == null) {
                    sig.label
                } else {
                    // TODO: Figure out how to underline the active parameter in addition to
                    //  bolding it to make it match the look of the signature help.
                    sig.label.substring(0, param.labelStart) +
                            "__" +
                            sig.label.substring(param.labelStart, param.labelEnd) +
                            "__" +
                            sig.label.substring(param.labelEnd)
                }
            )

            // Put the parameter doc and the instruction doc in the same string to match the look of the
            // signature help.
            var doc = ""

            // Parameter doc.
            if (param?.documentation != null) {
                doc += param.documentation

                // TODO: Figure out how add an empty line here to make it match the look of the
                //  signature help.
                doc += "\n\n"
            }

            // Instruction doc.
            sig.documentation?.let { doc += it }

            if (doc.isNotEmpty()) {
                contents.add(doc)
            }

            Hover(contents)
        }

        sendMessage(Response.GetHover(requestId, hover))
    }

    private fun getDefinition(requestId: Int, lineNo: Int, col: Int) {
        var result = emptyList<AsmRange>()

        getInstruction(lineNo, col)?.let { inst ->
            loop@
            for ((paramIdx, param) in inst.opcode.params.withIndex()) {
                if (param.type is LabelType) {
                    if (inst.opcode.stack != StackInteraction.Pop) {
                        // Immediate arguments.
                        val args = inst.getArgs(paramIdx)
                        val argSrcLocs = inst.getArgSrcLocs(paramIdx)

                        for (i in 0 until min(args.size, argSrcLocs.size)) {
                            val arg = args[i]
                            val srcLoc = argSrcLocs[i]

                            if (positionInside(lineNo, col, srcLoc)) {
                                val label = arg.value as Int
                                result = getLabelDefinitions(label)
                                break@loop
                            }
                        }
                    } else {
                        // Stack arguments.
                        val argSrcLocs = inst.getStackArgSrcLocs(paramIdx)

                        for ((i, argSrcLoc) in argSrcLocs.withIndex()) {
                            if (positionInside(lineNo, col, argSrcLoc)) {
                                val labelValues = getStackValue(cfg, inst, argSrcLocs.lastIndex - i)

                                if (labelValues.size <= 5) {
                                    result = labelValues.flatMap(::getLabelDefinitions)
                                }

                                break@loop
                            }
                        }
                    }
                }
            }
        }

        sendMessage(Response.GetDefinition(requestId, result))
    }

    private fun getInstruction(lineNo: Int, col: Int): Instruction? {
        for (segment in bytecodeIr.segments) {
            if (segment is InstructionSegment) {
                // Loop over instructions in reverse order so stack popping instructions will be
                // handled before the related stack pushing instructions when inlineStackArgs is on.
                for (i in segment.instructions.lastIndex downTo 0) {
                    val inst = segment.instructions[i]

                    inst.srcLoc?.let { srcLoc ->
                        if (positionInside(lineNo, col, srcLoc.mnemonic)) {
                            return inst
                        }

                        for (argSrcLoc in srcLoc.args) {
                            if (positionInside(lineNo, col, argSrcLoc)) {
                                return inst
                            }
                        }

                        if (inlineStackArgs) {
                            for (argSrcLoc in srcLoc.stackArgs) {
                                if (positionInside(lineNo, col, argSrcLoc)) {
                                    return inst
                                }
                            }
                        }
                    }
                }
            }
        }

        return null
    }

    private fun getLabelDefinitions(label: Int): List<AsmRange> =
        bytecodeIr.segments.asSequence()
            .filter { label in it.labels }
            .mapNotNull { segment ->
                val labelIdx = segment.labels.indexOf(label)

                segment.srcLoc.labels.getOrNull(labelIdx)?.let { labelSrcLoc ->
                    AsmRange(
                        startLineNo = labelSrcLoc.lineNo,
                        startCol = labelSrcLoc.col,
                        endLineNo = labelSrcLoc.lineNo,
                        endCol = labelSrcLoc.col + labelSrcLoc.len,
                    )
                }
            }
            .toList()

    private fun positionInside(lineNo: Int, col: Int, srcLoc: SrcLoc?): Boolean =
        if (srcLoc == null) {
            false
        } else {
            lineNo == srcLoc.lineNo && col >= srcLoc.col && col <= srcLoc.col + srcLoc.len
        }

    @Suppress("RedundantNullableReturnType") // Can return undefined.
    private fun getLine(lineNo: Int): String? = asm[lineNo - 1]

    companion object {
        private val KEYWORD_REGEX = Regex("""^\s*\.[a-z]+${'$'}""")
        private val KEYWORD_SUGGESTIONS: List<CompletionItem> =
            listOf(
                CompletionItem(
                    label = ".code",
                    type = CompletionItemType.Keyword,
                    insertText = "code",
                ),
                CompletionItem(
                    label = ".data",
                    type = CompletionItemType.Keyword,
                    insertText = "data",
                ),
                CompletionItem(
                    label = ".string",
                    type = CompletionItemType.Keyword,
                    insertText = "string",
                ),
            )

        private val INSTRUCTION_REGEX = Regex("""^\s*([a-z][a-z0-9_=<>!]*)?${'$'}""")
        private val INSTRUCTION_SUGGESTIONS: List<CompletionItem> =
            (OPCODES.asSequence() + OPCODES_F8.asSequence() + OPCODES_F9.asSequence())
                .filterNotNull()
                .map { opcode ->
                    CompletionItem(
                        label = opcode.mnemonic,
                        // TODO: Add signature?
                        type = CompletionItemType.Opcode,
                        insertText = opcode.mnemonic,
                    )
                }
                .sortedBy { it.label }
                .toList()
    }
}
